syntax = "proto3";

import "google/protobuf/empty.proto";
package graph;

service Graph {
  rpc AddHost(Host) returns (google.protobuf.Empty) {}
  rpc AddVertex(stream Vertex) returns (GraphSummary) {}
  rpc DeleteVertex(stream Vertex) returns (GraphSummary) {}
  rpc AddEdge(stream Edge) returns (GraphSummary) {}
  rpc DeleteEdge(stream Edge) returns (GraphSummary) {}
  rpc Search(SearchArgs) returns (SearchResults) {}
  rpc Ping(PingRequest) returns (PingResponse) {}
}

message Host {
  string key = 1;
  string address = 2;
}

message Vertex {
  string key = 1;
  string value = 2;
}

message Edge {
  string from = 1;
  string to = 2;
  string label = 3;
  string lookup_from = 4;
  string lookup_to = 5;
  string key = 6;
}

message GraphSummary {
  int32 vertex_count = 1;
  int32 edge_count = 2;
}

message SearchArgs {
  string start_key = 1;
  int32 level = 2;
  repeated Vertex vertices = 3;
  repeated Edge edges = 4;
  repeated string ids_so_far = 5;
}

message SearchResults {
  repeated Vertex vertices = 1;
  repeated Edge edges = 2;
  repeated string ids_so_far = 3;
}

message PingRequest {
  string data = 1;
}

message PingResponse {
  string data = 1;
}